assets/environment-interpreter/activate/activate.d/zsh

```diff
@@ -43,21 +43,39 @@ done
 if [ ${#fpath_prepend[@]} -gt 0 ]; then
   fpath=( "${fpath_prepend[@]}" "${fpath[@]}" )
   autoload -U compinit
+  declare -a _flox_compinit_args=()
 
-  # compinit tracing is very verbose so only enable with tracelevel 3 or above.
+  # Log the reason for any recompilations to STDERR at tracelevel >=1.
+  # This flag was introduced in 5.9 but older versions will just ignore it.
+  if [ "$_flox_activate_tracelevel" -ge 1 ]; then
+    _flox_compinit_args+=( "-w" )
+  fi
+
+  # compinit tracing is very verbose so only enable at tracelevel >=3
+  #
+  # We only need to explicitly handle tracelevel =2 because:
+  # - <2 already has tracing disabled
+  # - >3 always has tracing enabled
   if [ "$_flox_activate_tracelevel" -eq 2 ]; then
-    # Disable tracing for the compinit invocation then re-enable.
     : compinit tracing suppressed
     set +x
-    compinit
-    set -x
-  else
-    # Tracing will either have already been enabled for tracelevels greater or
-    # equal to 2 or not enabled for tracelevels below 2, but either way we can
-    # leave tracing as-is and just invoke compinit.
-    compinit
   fi
 
+  # Cache completions where possible.
+  #
+  # `activate` doesn't require `--env-cache` so we have to be careful that it's
+  # set. `compinit -d` persists the value as `$_comp_dumpfile` in the shell so
+  # it may write to the wrong file if the environment doesn't have `--env-cache`
+  if [[ -n "${FLOX_ENV_CACHE:-}" && -d "${FLOX_ENV_CACHE}" ]]; then
```

Comment:
Can you write a little more about what gets cached when and where? Without doing my own research into `compinit`, it's not clear to me in what scenarios the completion cache will be regenerated when caching in `FLOX_ENV_CACHE`.

I've read the PR description, but it would also be nice to have this documented in the code as well.

Reply:
Nerd sniped:
- `flox activate <args>` calls the activation script with `--env-cache <prefix>/.flox/cache`, regardless of which other environments are already active.
- At activation time, you source this `zsh` script, which loads completions via `compinit`.
- `compinit` loads completions from directories in `$fpath`, and uses the number of completion files to determine whether to reinitialize.
- That means changes to an individual environment, or activation with a different number/set of layered environments compared to the most recent activation will cause a cache miss.

Reply:
With regards to the other caching mechanisms you mentioned in the PR description, I also don't think they're worth the effort at the moment. You'd need some kind of garbage collection mechanism, otherwise you'd just keep growing a list of hashed `FLOX_ENV_DIRS` values indefinitely, and that's well and truly outside the scope of the ticket.